A51-custom-backend-metrics.md

```diff
@@ -279,67 +303,86 @@ public final class CallMetricRecorder {
   // Records a utilization metric measurement for the call.
   public CallMetricRecorder recordUtilizationMetric(String name, double value);
 
-  //Records the CPU utilization metric measurement for the call. 
+  //Records the CPU utilization metric measurement for the call.
   public CallMetricRecorder recordCPUUtilizationMetric(double value);
 
-  // Records the memory utilization metric measurement for the call. 
+  // Records the memory utilization metric measurement for the call.
   public CallMetricRecorder recordMemoryUtilizationMetric(double value);
 
+  // Records the queries per second measurement.
+  public CallMetricRecorder recordQpsMetric(double value);
+
   // Returns the call metric recorder attached to the current call.
   public static CallMetricRecorder getCurrent();
 }
 ```
 
-At the client side, Java will provide a utility function `OrcaClientStreamTracerFactory`. 
-Users can include it in the picker result in their custom LB policy’s 
-`LoadBalancer.SubchannelPicker` to let gRPC push notifications to the application.
-It requires users to implement `OrcaPerRequestReportListener` in order to receive the metrics report callback.
-To avoid double-parsing trailers, the utility will use a CallOption in StreamInfo.
-If the call option is not already present, the utility will create an object to parse the trailer
-and notify listeners. The call option is then added and references that created object. If the call
-option is already present, the listener will be added to the list of listeners to be notified.
-
-```Java
-class CustomPicker extends SubchannelPicker {
-  public PickResult pickSubchannel(PickSubchannelArgs args) {
-  Subchannel subchannel = … // WRR picking logic
-  return PickResult.withSubchannel(
-    subchannel,
-    OrcaPerRequestUtil.newOrcaClientStreamTracerFactory(listener));
-  }
-}
-```
-
-#### Out-of-Band Reporting 
-At the server side, Java implemented the OOB service `OrcaServiceImpl`: users can optionally provide the minimum
-reporting interval configuration when creating the object. Then they can update utilization metrics
-data using the exposed APIs as follows. All the methods are thread safe.
+#### Per-Server Recording on Server
+Java will provide a per-server metrics recorder on the server as follows. The user
+creates an instance and records metrics to it. All the methods are thread safe.
+This is optional for per-request load reporting and required for OOB load reporting.
 
 ```java
 public class MetricRecorder {
   // Update the metrics value corresponding to the specified key.
   public void putUtilizationMetric(String key, double value);
 
   // Replace the whole metrics data using the specified map.
-  public void setAllUtilizationMetrics(Map<String, Double> metrics);
+  public void putAllUtilizationMetrics(Map<String, Double> metrics);
```

Comment:
Don't change the Java API... These are sets, and the qps metric should also be a set.

Reply:
Reverted. FYI left `putUtilizationMetric` as is.